lib/wallet/src/Cardano/Wallet/Write/Tx/Gen.hs

```diff
@@ -0,0 +1,105 @@
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Copyright: © 2022 IOHK
+-- License: Apache-2.0
+--
+module Cardano.Wallet.Write.Tx.Gen
+    ( genDatum
+    , genData
+    , genDatumHash
+    , shrinkData
+    , shrinkDatum
+    , genTxOut
+    )
+    where
+
+import Prelude
+
+import Cardano.Ledger.Alonzo.Data
+    ( Data (..), dataToBinaryData )
+import Cardano.Wallet.Write.Tx
+    ( BinaryData
+    , Datum (..)
+    , DatumHash
+    , LatestLedgerEra
+    , RecentEra
+    , ShelleyLedgerEra
+    , TxOut
+    , cardanoEraFromRecentEra
+    , datumFromCardanoScriptData
+    , datumHashFromBytes
+    , datumToCardanoScriptData
+    , shelleyBasedEraFromRecentEra
+    )
+import Data.ByteString
+    ( ByteString )
+import Data.Maybe
+    ( fromMaybe )
+import Ouroboros.Consensus.Cardano.Block
+    ( EraCrypto, StandardCrypto )
+import Test.QuickCheck
+    ( Gen
+    , arbitrary
+    , choose
+    , listOf
+    , oneof
+    , scale
+    , shrinkMapBy
+    , sized
+    , vector
+    , vectorOf
+    )
+
+import qualified Cardano.Api.Gen as Cardano
+import qualified Cardano.Api.Shelley as Cardano
+import qualified Data.ByteString as BS
+import qualified Plutus.V1.Ledger.Api as PV1
+
+genDatum :: (EraCrypto era ~ StandardCrypto) => Gen (Datum era)
+genDatum = oneof
+    [ Datum <$> genData
+    , DatumHash <$> genDatumHash
+    , pure NoDatum
+    ]
+
+-- Originally from https://github.com/input-output-hk/cardano-ledger/blob/c7c63dabdb215ebdaed8b63274965966f2bf408f/eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Serialisation/Generators.hs#L66-L79
+genData :: Gen (BinaryData era)
+genData = dataToBinaryData . Data <$> scale (min 7) (sized gendata)
```

Comment:
After some consideration and experimentation, I think it might make sense.

I was concerned over too large `BinaryData` values accidentally being generated, and was also not sure what the domain of sizes would be if we used `div`. 

But the sizes seem to range from 0 to 99

```haskell
quickCheck $ withMaxSuccess 10000 $ forAll getSize $ \s -> tabulate "size" [show s] $ property True

+++ OK, passed 10000 tests.

size (10000 in total):
 5.17% 0
 ...
 0.03% 99
```

where `scale (min 6)` would approximate `scale (const 6)`. So I guess something like `flip div 10` would allow us to have smaller values on average, without sacrificing extreme values.

Reply:
430a26a89d3f45835f296b32c27c1c20e09b3f36

Reply:
I'm not sure a `scale (2 *)` of a wrapping datatype (e.g. TxOut) would actually want to affect `genBinaryData` in its current form... but that's a problem for if/then.